operators/src/processing/expression/query_processor.rs

```diff
@@ -172,38 +174,46 @@ where
     }
 
     #[inline]
+    #[allow(clippy::eq_op)]
+    #[allow(clippy::float_cmp)]
     fn compute_expression(
         raster: Self::Tuple,
         program: &LinkedExpression,
         map_no_data: bool,
         out_no_data: TO,
-    ) -> Result<Vec<TO>> {
+    ) -> Result<(Vec<TO>, Vec<bool>)> {
         let expression = unsafe {
             // we have to "trust" that the function has the signature we expect
             program.function_3::<f64, bool, f64>()?
         };
 
         // cannot be empty at this point
         let tile = raster.into_materialized_tile();
+        let x_axis_size = tile.grid_array.inner_grid.grid_shape().axis_size_x();
 
-        let data = tile
-            .grid_array
-            .data
+        let res = (
+            tile.grid_array.inner_grid.data,
+            tile.grid_array.validity_mask.data,
+        )
             .par_iter()
-            .with_min_len(tile.grid_array.grid_shape().axis_size_x())
-            .map(|a| {
-                let is_no_data = tile.is_no_data(*a);
-
-                if !map_no_data && is_no_data {
-                    return out_no_data;
+            .with_min_len(x_axis_size)
+            .map(|(a, &is_a_valid)| {
+                if !map_no_data && !is_a_valid {
+                    return (out_no_data, false);
                 }
 
-                let result = expression(a.as_(), is_no_data, out_no_data.as_());
-                TO::from_(result)
+                let result = expression(a.as_(), !is_a_valid, out_no_data.as_());
+
+                let result_t = TO::from_(result);
+
+                let out_is_no_data =
+                    result_t == out_no_data || (out_no_data != out_no_data && result != result);
```

Comment:
what is happening here?

Reply:
we need to check if the result value is the no data value the user set. so the result is either equal to the no_data value or both result and no_data are NAN.